scripts/sentiment_analysis/process_data.py

```diff
@@ -0,0 +1,138 @@
+        return string.strip().lower()
+    else:
+        string = re.sub(r'[^A-Za-z0-9(),!?\'\`]', ' ', string)
+        string = re.sub(r'\'s', ' \'s', string)
+        string = re.sub(r'\'ve', ' \'ve', string)
+        string = re.sub(r'n\'t', ' n\'t', string)
+        string = re.sub(r'\'re', ' \'re', string)
+        string = re.sub(r'\'d', ' \'d', string)
+        string = re.sub(r'\'ll', ' \'ll', string)
+        string = re.sub(r',', ' , ', string)
+        string = re.sub(r'!', ' ! ', string)
+        string = re.sub(r'\(', ' ( ', string)
+        string = re.sub(r'\)', ' ) ', string)
+        string = re.sub(r'\?', ' ? ', string)
+        string = re.sub(r'\s{2,}', ' ', string)
+        return string.strip() if data_name == 'TREC' else string.strip().lower()
+
+def _build_vocab(data_name, train_dataset, test_dataset):
+    all_token = []
+    max_len = 0
+    for i, line in enumerate(train_dataset):
+        train_dataset[i][0] = _clean_str(line[0], data_name)
+        line = train_dataset[i][0].split()
+        max_len = max_len if max_len > len(line) else len(line)
+        all_token.extend(line)
+    for i, line in enumerate(test_dataset):
+        test_dataset[i][0] = _clean_str(line[0], data_name)
+        line = test_dataset[i][0].split()
+        max_len = max_len if max_len > len(line) else len(line)
+    all_token.extend(line)
+    vocab = nlp.Vocab(nlp.data.count_tokens(all_token))
+    vocab.set_embedding(nlp.embedding.create('Word2Vec', source='GoogleNews-vectors-negative300'))
+    for line in vocab.embedding._idx_to_token:
+        if (vocab.embedding[line] == nd.zeros(300)).sum() == 300:
+            vocab.embedding[line] = nd.random.normal(-1.0, 1.0, 300)
+    vocab.embedding['<unk>'] = nd.zeros(300)
+    vocab.embedding['<pad>'] = nd.zeros(300)
+    vocab.embedding['<bos>'] = nd.zeros(300)
+    vocab.embedding['<eos>'] = nd.zeros(300)
+    print('maximum length (in tokens): ', max_len)
+    return vocab, max_len
+
+# Dataset preprocessing.
+def _preprocess(x, vocab, max_len):
+    data, label = x
+    data = vocab[data.split()]
+    if len(data) > max_len:
+        data = data[:max_len]
+    else:
+        while len(data) < max_len:
+            data.append(0)
+    return data, label
+
+def _get_length(x):
+    return float(len(x[0]))
+
+def _preprocess_dataset(dataset, vocab, max_len):
+    start = time.time()
+    pool = mp.Pool(8)
+    partial_work = partial(_preprocess, vocab=vocab, max_len=max_len)
+    dataset = pool.map(partial_work, dataset)
+    lengths = gluon.data.SimpleDataset(pool.map(_get_length, dataset))
+    end = time.time()
+    print('Done! Tokenizing Time={:.2f}s, #Sentences={}'.format(end - start, len(dataset)))
+    return dataset, lengths
+
+def load_dataset(data_name):
+    """Load sentiment dataset."""
+    if data_name == 'MR' or data_name == 'Subj':
+        train_dataset, output_size = _load_file(data_name)
+        vocab, max_len = _build_vocab(data_name, train_dataset, [])
```

Comment:
Arent we truncating ?
like max(1000, np.percentile(lenghts,99)) etc just to safegaurd us ?

Reply:
These are according to experimental settings of the textCNN paper/implementations.